src/promptflow-evals/promptflow/evals/evaluators/chat/__init__.py

```diff
@@ -0,0 +1,189 @@
+            role = each_turn["role"]
+            if role == "user":
+                questions.append(each_turn["content"])
+            elif role == "assistant":
+                answers.append(each_turn["content"])
+                if "context" in each_turn and "citations" in each_turn["context"]:
+                    citations = json.dumps(each_turn["context"]["citations"])
+                    contexts.append(citations)
+
+        # Select evaluators to be used for evaluation
+        compute_rag_based_metrics = True
+        if len(answers) != len(contexts):
+            safe_message = "Skipping rag based metrics as we need citations or " \
+                            "retrieved_documents in context key of every assistant's turn"
+            logger.warning(safe_message)
+            compute_rag_based_metrics = False
+
+        selected_evaluators = []
+        selected_evaluators.extend(self._non_rag_evaluators)
+        if compute_rag_based_metrics:
+            selected_evaluators.extend(self._rag_evaluators)
+
+        # Evaluate each turn
+        per_turn_results = []
+        for turn_num in range(len(questions)):
+            current_turn_result = {}
+
+            if parallel:
+                # Parallel execution
+                with ThreadPoolExecutor() as executor:
+                    future_to_evaluator = {
+                        executor.submit(self._evaluate_turn, turn_num, questions, answers, contexts, evaluator): evaluator
+                        for evaluator in selected_evaluators
+                    }
+
+                    for future in as_completed(future_to_evaluator):
+                        score = future.result()
+                        current_turn_result.update(score)
+            else:
+                # Sequential execution
+                for evaluator in selected_evaluators:
+                    score = self._evaluate_turn(turn_num, questions, answers, contexts, evaluator)
+                    current_turn_result.update(score)
+
+            per_turn_results.append(current_turn_result)
+
+        # Aggregate results
+        # Final aggregated results for a conversation will look like:
+        # {
+        #     "gpt_groundedness": 0.9,
+        #     "gpt_groundedness_per_turn": [0.9, 0.8, 0.9, ...],
+        #     ...
+        # }
+        aggregated = {}
+        for key in per_turn_results[0].keys():
+            values = [d[key] for d in per_turn_results]
+            aggregated[key] = np.nanmean(values)
+            aggregated[key + "_per_turn"] = values
+
+        return aggregated
+
+    def _evaluate_turn(self, turn_num, questions, answers, contexts, evaluator):
+        try:
+            question = questions[turn_num] if turn_num < len(questions) else ""
+            answer = answers[turn_num] if turn_num < len(answers) else ""
+            context = contexts[turn_num] if turn_num < len(contexts) else ""
+
+            score = evaluator(
```

Comment:
Does individual evaluator handle is case say answer/question/context is empty ?

Reply:
Correct. Every individual evaluator has its own input validation logic. 
Groundedness for example:
https://github.com/microsoft/promptflow/blob/users/singankit/promptflow-eval/src/promptflow-evals/promptflow/evals/evaluators/groundedness/flow/validate_inputs.py